packages/node-runtime-worker-thread/src/serializer.ts

```diff
@@ -0,0 +1,145 @@
+import v8 from 'v8';
+import { ShellResult } from '@mongosh/shell-evaluator';
+import { inspect } from 'util';
+import { EJSON, Document } from 'bson';
+
+export function serialize(data: unknown): string {
+  return `data:;base64,${v8.serialize(data).toString('base64')}`;
+}
+
+export function deserialize<T = unknown>(str: string): T | string {
+  if (/^data:;base64,.+/.test(str)) {
+    return v8.deserialize(
+      Buffer.from(str.replace('data:;base64,', ''), 'base64')
+    );
+  }
+  return str;
+}
+
+function isPrimitive(
+  val: any
+): val is boolean | string | number | undefined | null {
+  return (typeof val !== 'object' && typeof val !== 'function') || val === null;
+}
+
+function isError(val: any): val is Error {
+  return val && val.name && val.message && val.stack;
+}
+
+function getNames<T>(obj: T): (keyof T)[] {
+  return Object.getOwnPropertyNames(obj) as (keyof T)[];
+}
+
+/**
+ * Extracts non-enumerable params from errors so they can be serialized and
+ * de-serialized when passed between threads.
+ *
+ * Even though v8 {de}serialize supports Error serialization, some data (e.g.,
+ * error `name` if you have a custom error) still can be lost during this
+ * process, so serializing it produces a better output.
+ */
+export function serializeError(err: Error) {
+  // Name is the only constructor property we care about
+  const keys = getNames(err).concat('name');
+  return keys.reduce((acc, key) => {
+    (acc as any)[key] = err[key];
+    return acc;
+  }, {} as Error);
+}
+
+/**
+ * Creates an instance of Error from error params (Error-like objects)
+ */
+export function deserializeError(err: any): Error {
+  return Object.assign(new Error(), err);
+}
+
+export enum SerializedResultTypes {
+  SerializedErrorResult = 'SerializedErrorResult',
+  InspectResult = 'InspectResult',
+  SerializedShellApiResult = 'SerializedShellApiResult'
+}
+
+type SerializedShellApiResult = {
+  origType: string;
+  serializedValue: Document;
+  nonEnumPropertyDescriptors: Record<string, PropertyDescriptor>;
+};
+
+export function serializeEvaluationResult(result: ShellResult): ShellResult {
+  result = { ...result };
+
+  // Type `null` indicates anything that is not returned by shell-api, any
+  // usual javascript value returned from evaluation
+  if (result.type === null) {
```

Comment:
Wow going through all these nuances must have been though! :)

What do you think about making all of these ifs go away with early returns and get rid of mutations ?

As a reader I found quite hard to make sense of all these corner case, I'm thinking a maintainer without much context could struggle to be able to confidently change and predict the outcome of this crucial code.

It could help to have less necessity for so many comments too.

For example, for what i get from the code if either `isPrimitive(result.printable) or isPrimitive(result.rawValue)` it seems that nothing is happening... in that case and the result stays almost untouched. We can take care of that case immediately and relief the mind of the reader from that.

Same would go with an error i don't thing is needed for that to be in a nested if. The remaining part of the function would be for the 'main' happy path:

``` js
const { type, rawValue, printable } = result;

if (isPrimitive(result.rawValue) || isPrimitive(result.printable)) {
   return { type, printable };
}

if (isError(result.rawValue)) {
  return { 
  ...result,
   type: SerializedResultTypes.SerializedErrorResult,
   printable: serializeError(result.rawValue)
 }
}

return {
  type: SerializedResultTypes.SerializedShellApiResult,
  printable: {
    origType: result.type,
    serializedValue: EJSON.serialize(result.printable),
    nonEnumPropertyDescriptors: Object.fromEntries(
      Object.entries(
        Object.getOwnPropertyDescriptors(result.printable)
      ).filter(([, descriptor]) => !descriptor.enumerable)
    )
  }
}
```

Reply:
Definitely helps to have a fresh look at the code here, thanks! Yeah, totally happy to simplify it as you suggested. I'd prefer to keep the comments though if you don't mind, I think even with less nesting, it's still valuable to have some additional context on what exactly we are doing here and what those checks imply in scope of `ShellResult`s beyond the literal logic of the code there

Reply:
Done in e42b23b